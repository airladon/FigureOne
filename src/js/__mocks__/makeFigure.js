// import {
//   FigureElementPrimitive,
//   FigureElementCollection,
//   // AnimationPhase,
// } from '../Element';
import Figure from '../figure/Figure';
import {
  Rect, getPoint, Point, getRect,
} from '../tools/g2';
import webgl from './WebGLInstanceMock';
import DrawContext2D from './DrawContext2DMock';
import * as tools from '../tools/tools';
import { round } from '../tools/math';

tools.isTouchDevice = jest.fn();

jest.mock('../figure/Gesture');
jest.mock('../figure/webgl/webgl');
jest.mock('../figure/DrawContext2D');
jest.mock('../figure/DrawingObjects/HTMLObject/HTMLObject');

const generateRandomStringMock = jest.spyOn(tools, 'generateRandomString');
generateRandomStringMock.mockImplementation(() => '000000');

export default function makeFigure(
  inputCanvasIn = new Rect(100, -300, 1000, 500),
  scene = {},
) {
  document.body.innerHTML =
    '<div id="c">'
    + '  <canvas class="figureone__gl" id="id_figureone__gl__low">'
    + '  </canvas>'
    + '  <canvas class="figureone__text" id="id_figureone__text__low">'
    + '  </canvas>'
    + '  <div class="figureone__html">'
    + '  </div>'
    + '  <canvas class="figureone__gl" id="id_figureone__gl__high">'
    + '  </canvas>'
    + '  <canvas class="figureone__text" id="id_figureone__text__high">'
    + '  </canvas>'
    + '</div>';
  // canvas = document.getElementById('c');
  const inputCanvas = getRect(inputCanvasIn);
  // const inputLimits = getRect(inputLimitsIn);
  const definition = {
    width: inputCanvas.width,
    height: inputCanvas.height,
    limits: new Rect(-1, -1, 2, 2),
  };

  const canvasMock = {
    width: definition.width,
    height: definition.height,
    // offsetLeft: 100,
    left: inputCanvas.left,
    // offsetTop: 200,
    top: inputCanvas.top,
    // width: definition.width,
    // height: definition.height,
    offsetWidth: definition.width,
    offsetHeight: definition.height,
    scrollLeft: 0,
    scrollTop: 0,
    // eslint-disable-next-line arrow-body-style
    getBoundingClientRect: () => {
      return {
        left: inputCanvas.left,
        top: inputCanvas.top,
        width: definition.width,
        height: definition.height,
      };
    },
  };
  const htmlCanvasMock = {
    style: {
      fontsize: 1,
    },
    offsetWidth: 100,
    appendChild: () => {},
  };
  const { limits } = definition;
  const figure = new Figure({
    htmlId: 'c', limits, color: [1, 0, 0, 1], scene,
  });
  figure.webglLow = webgl;
  figure.webglHigh = webgl;
  figure.webgl = webgl;
  figure.canvasLow = canvasMock;
  figure.webglLow.gl.canvas = canvasMock;
  figure.canvasHigh = canvasMock;
  figure.htmlCanvas = htmlCanvasMock;
  figure.isTouchDevice = false;
  figure.draw2DLow = new DrawContext2D(definition.width, definition.height);
  figure.draw2DHigh = new DrawContext2D(definition.width, definition.height);
  figure.draw2D = figure.draw2DLow;
  figure.shapesLow = figure.getShapes(false);
  figure.shapesHigh = figure.getShapes(true);
  figure.shapes = figure.shapesLow;
  figure.primitives = figure.shapes;
  // figure.equationLow = figure.getEquations(false);
  // figure.equationHigh = figure.getEquations(true);
  figure.equation = figure.equationLow;
  figure.collectionsLow = figure.getObjects(false);
  figure.collectionsHigh = figure.getObjects(true);
  figure.collections = figure.collectionsLow;
  figure.createFigureElements();
  figure.defaultLineWidth = 0.01;
  figure.primitives.defaultLineWidth = 0.01;
  // needed as the first element needs to be set with the space Transforms
  figure.initElements();
  figure.mock = {
    initialTime: 0,
    duration: 0,
    previousTouchPoint: new Point(0, 0),
    timersBeforeDraw: true,
    timeStep: (deltaTimeInSecondsIn, frameTimeIn = null, draw = true) => {
      const { duration, initialTime } = figure.mock;
      const deltaTimeInSeconds = round(deltaTimeInSecondsIn, 8);
      let frameTime = deltaTimeInSeconds;
      // let deltaTime = 0;
      if (frameTimeIn != null && frameTimeIn < deltaTimeInSeconds) {
        frameTime = frameTimeIn;
        // deltaTime = frameTime;
      }
      let deltaTime = frameTime;

      // let delta = Math.min(deltaTimeInSeconds, frameTime);
      let lastTime = 0;
      deltaTime = round(deltaTime, 8);
      while (deltaTime <= round(deltaTimeInSeconds, 8)) {
        const newNow = round((duration + deltaTime + initialTime) * 1000, 8);
        global.performance.now = () => newNow;
        if (figure.mock.timersBeforeDraw) {
          jest.advanceTimersByTime(round((deltaTime - lastTime) * 1000, 8));
        }
        if (draw) {
          figure.draw(round(newNow / 1000, 8));
        }
        if (!figure.mock.timersBeforeDraw) {
          jest.advanceTimersByTime(round((deltaTime - lastTime) * 1000, 8));
        }
        lastTime = deltaTime;
        if (deltaTime === deltaTimeInSeconds || frameTime === 0) {
          deltaTime += 1;
        } else if (deltaTime + frameTime > deltaTimeInSeconds) {
          deltaTime = deltaTimeInSeconds;
        } else {
          deltaTime += frameTime;
        }
        deltaTime = round(deltaTime, 8);
      }
      figure.mock.duration += deltaTimeInSeconds;
    },
    touchDown: (figurePosition) => {
      const p = getPoint(figurePosition);
      figure.touchDownHandler(p);
      figure.mock.previousTouchPoint = p;
    },
    touchUp: () => {
      figure.touchUpHandler();
    },
    touchMove: (figurePosition) => {
      const p = getPoint(figurePosition);
      figure.touchMoveHandler(figure.mock.previousTouchPoint, p);
      figure.mock.previousTouchPoint = p;
    },
  };
  figure.timeKeeper.reset();
  return figure;
}

