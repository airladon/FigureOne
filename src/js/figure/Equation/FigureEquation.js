// @flow
import WebGLInstance from '../webgl/webgl';
import {
  Rect,
} from '../../tools/g2';
import {
  FigureElementCollection,
} from '../Element';
import { joinObjects } from '../../tools/tools';
import DrawContext2D from '../DrawContext2D';
import { Equation } from './Equation';
import type { EQN_Equation } from './Equation';
import type { TypeNavigatorOptions } from '../FigureCollections/EqnNavigator';
import EqnNavigator from '../FigureCollections/EqnNavigator';

export type TypeEquationPhrase =
  string
  | Array<string
    | TypeEquationPhrase
    | {                                 // Fraction and SmallFraction
      numerator: TypeEquationPhrase;
      denominator: TypeEquationPhrase;
      symbol: string;
      scaleModifier?: number;
    }
    | {                                 // Superscript and Subscript
      content: TypeEquationPhrase;
      subscript?: TypeEquationPhrase;
      superscript?: TypeEquationPhrase;
    }
    | {                                 // Strike
      content: TypeEquationPhrase;
      symbol: string;
      inSize?: boolean;
    }
    | {                                 // Annotation
      content: TypeEquationPhrase;
      annotationArray: Array<{
        content: TypeEquationPhrase;
        xPosition?: 'left' | 'right' | 'center';
        yPosition?: 'bottom' | 'top' | 'middle' | 'baseline';
        xAlign?: 'left' | 'right' | 'center';
        yAlign?: 'bottom' | 'top' | 'middle' | 'baseline';
        scale?: number;
      }>;
      inSize?: boolean;
    }
    | {                                 // Bracket
      content: TypeEquationPhrase;
      symbol: string;
      symbol: string;
      space?: number;
    }
    | {                                 // Top and Bottom Bar
      content: TypeEquationPhrase;
      symbol: string;
      space?: number;
      outsideSpace?: number;
    }
    | {                                 // Top and Bottom Comment
      content: TypeEquationPhrase;
      comment: TypeEquationPhrase;
      symbol: string;
      space?: number;
      outsideSpace?: number;
    }>;

export default class FigureEquation {
  webgl: Array<WebGLInstance>;
  draw2D: DrawContext2D;
  limits: Rect;
  shapes: Object;
  animateNextFrame: () => void;

  constructor(
    shapes: Object,
    animateNextFrame: () => void,
  ) {
    this.webgl = shapes.webgl;
    this.draw2D = shapes.draw2D;
    this.limits = shapes.limits;
    this.shapes = shapes;
    this.animateNextFrame = animateNextFrame;
  }

  equation(
    options: EQN_Equation,
  ) {
    const equation = new Equation(this.shapes, options);
    return equation;
  }

  addEquation(
    parent: FigureElementCollection,
    name: string,
    options: EQN_Equation = {},
  ) {
    // $FlowFixMe
    const equation = new Equation(this.shapes, options);
    parent.add(name, equation);
    return equation;
  }

  addNavigator(
    parent: FigureElementCollection,
    name: string,
    options: TypeNavigatorOptions,
  ) {
    let navNameToUse = name;
    const optionsToUse = joinObjects({}, options);
    if (optionsToUse.equation == null) {                // $FlowFixMe
      const equation = this.addEquation(parent, `${name}Eqn`, options);
      optionsToUse.equation = equation;
      navNameToUse = `${name}Nav`;
    } else if (!(optionsToUse.equation instanceof Equation)) {
      // let methodPathToUse;
      let nameToUse;
      // let pathToUse;
      let eqnOptions;
      let elementModsToUse;
      // let addElementsToUse;
      let firstScenario;
      if (Array.isArray(optionsToUse.equation)) {
        [, nameToUse, , eqnOptions,
          elementModsToUse, , firstScenario,
        ] = optionsToUse.equation;
      } else {
        nameToUse = optionsToUse.equation.name;
        // pathToUse = optionsToUse.equation.path;
        eqnOptions = optionsToUse.equation.options;
        // methodPathToUse = optionsToUse.equation.method;
        elementModsToUse = optionsToUse.equation.mods;
        firstScenario = optionsToUse.equation.scenario;
      }

      let equation;
      if (Array.isArray(eqnOptions)) {
        equation = this.addEquation(parent, nameToUse, ...eqnOptions);
      } else {
        equation = this.addEquation(parent, nameToUse, eqnOptions);
      }

      if (elementModsToUse != null && elementModsToUse !== {}) {
        equation.setProperties(elementModsToUse);
      }
      if (firstScenario != null && firstScenario in equation.scenarios) {
        equation.setScenario(firstScenario);
      }
      optionsToUse.equation = equation;
    }
    // $FlowFixMe
    const navigator = new EqnNavigator(
      this.shapes,
      this.animateNextFrame,
      optionsToUse,
    );
    parent.add(navNameToUse, navigator);
    return navigator;
  }
}
