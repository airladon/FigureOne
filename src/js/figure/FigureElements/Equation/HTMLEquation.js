// @flow

function makeDiv(
  id: string,
  classes: Array<string>,
  text: string,
  indent: number = 0,
) {
  const indentStr = ' '.repeat(indent);
  const idStr = id ? ` id="${id}"` : '';
  const classString = classes ? ` ${classes.join(' ')}` : '';
  let out = `${indentStr}<div${idStr} class="equation_element${classString}">\n`;
  out += `${text}\n`;
  out += `${indentStr}</div>`;
  return out;
}

// Most fundamental Equation Element properties includes element size and
// location, as well as html id and classes.
class HTMLElementProperties {
  id: string;
  classes: Array<string>;

  constructor(id: string = '', classes: string | Array<string> = '') {
    this.id = id;
    if (Array.isArray(classes)) {
      this.classes = classes;
    } else if (classes.length > 0) {
      this.classes = classes.split(' ');
    } else {
      this.classes = [];
    }
  }

  render(indent: number = 0, text: string = '') {
    return makeDiv(
      this.id,
      this.classes,
      text,
      indent,
    );
  }
}

class HTMLElement extends HTMLElementProperties {
  text: string;

  constructor(
    text: string,
    id: string = '',
    classes: string | Array<string> = [],
  ) {
    super(id, classes);
    this.classes.push('equation_text');
    this.text = text;
  }

  render(indent: number = 0) {
    return super.render(indent, `${' '.repeat(indent + 2)}${this.text}`);
  }
}

class HTMLElements extends HTMLElementProperties {
  content: Array<HTMLElement | HTMLElements | HTMLElementProperties>;

  constructor(
    content: Array<HTMLElement | HTMLElements | HTMLElementProperties | null>,
    id: string = '',
    classes: string | Array<string> = [],
  ) {
    super(id, classes);
    const filteredContent = [];
    content.forEach((c) => {
      if (c !== null) {
        filteredContent.push(c);
      }
    });
    this.content = filteredContent;
  }

  render(indent: number = 0) {
    return super.render(indent, this.content.map(c => c.render(indent + 2)).join('\n'));
  }
}

class Fraction extends HTMLElementProperties {
  numerator: HTMLElements;
  denominator: HTMLElements;

  constructor(
    numerator: HTMLElements,
    denominator: HTMLElements,
    id: string = '',
    classes: string | Array<string> = [],
  ) {
    super(id, classes);
    this.classes.push('fraction');
    this.numerator = numerator;
    this.denominator = denominator;
  }

  render(indent: number = 0) {
    const s = ' '.repeat(indent + 2);
    let out = '';
    out += `${s}<div class="numerator">\n`;
    out += this.numerator.render(indent + 4);
    out += `\n${s}</div>\n`;
    out += `${s}<div class="fraction_line"> </div>\n`;
    out += `${s}<div class="denominator">\n`;
    out += this.denominator.render(indent + 4);
    out += `\n${s}</div>`;
    return super.render(indent, out);
  }
}

class SuperSub extends HTMLElementProperties {
  superscript: HTMLElements | null;
  subscript: HTMLElements | null;
  content: HTMLElements;

  constructor(
    content: HTMLElements,
    superscript: HTMLElements | null,
    subscript: HTMLElements | null,
    id: string = '',
    classes: string | Array<string> = '',
  ) {
    super(id, classes);
    this.classes.push('supersub');
    this.superscript = superscript;
    this.subscript = subscript;
    this.content = content;
  }

  render(indent: number = 0) {
    const s = ' '.repeat(indent + 2);
    let out = '';
    out += `${s}<div class="supsub_content element">\n`;
    out += this.content.render(indent + 2);
    out += `\n${s}</div>\n`;
    out += `${s}<div class="super_sub element">\n`;
    out += `${s}<div class="super_sub_super superscript_text element">\n`;
    if (this.superscript !== null) {
      out += this.superscript.render(indent + 4);
    }
    out += `\n${s}</div>\n`;
    out += `${s}<div class="super_sub_sub subscript_text element">\n`;
    if (this.subscript !== null) {
      out += this.subscript.render(indent + 4);
    }
    out += `\n${s}</div>`;
    out += `\n${s}</div>\n`;
    return super.render(indent, out);
  }
}

class Subscript extends SuperSub {
  constructor(
    content: HTMLElements,       // eslint-disable-line no-use-before-define
    subscript: HTMLElements,
    id: string = '',
    classes: string | Array<string> = '',
  ) {
    super(content, null, subscript, id, classes);
    const index = this.classes.indexOf('supersub');
    if (index > -1) {
      this.classes.splice(index, 1);
    }
    this.classes.push('subscript');
  }
}

class Superscript extends SuperSub {
  constructor(
    content: HTMLElements,       // eslint-disable-line no-use-before-define
    superscript: HTMLElements,
    id: string = '',
    classes: string | Array<string> = '',
  ) {
    super(content, superscript, null, id, classes);
    const index = this.classes.indexOf('supersub');
    if (index > -1) {
      this.classes.splice(index, 1);
    }
    this.classes.push('superscript');
  }
}


class Root extends HTMLElementProperties {
  content: HTMLElements;

  constructor(
    content: HTMLElements,     // eslint-disable-line no-use-before-define
    id: string = '',
    classes: string | Array<string> = [],
  ) {
    super(id, classes);
    this.content = content;
  }

  render(indent: number = 0) {
    const s = ' '.repeat(indent + 2);
    let out = '';
    out += `${s}<div class="square_root element">\n`;
    out += `${s}  <div class="equation_element radical element">\n`;
    out += `${s}    &radic;\n`;
    out += `${s}  </div>\n`;
    out += this.content.render(indent + 4);
    out += `\n${s}</div>`;
    return super.render(indent, out);
  }
}

// function contentToE(content: string | HTMLElements): HTMLElements {
//   let c;
//   if (typeof content === 'string') {
//     c = new HTMLElements(content);
//   } else {
//     c = content;
//   }
//   return c;
// }

type EquationInput = Array<HTMLElements | HTMLElementProperties |
                           HTMLElement | string> |
                     HTMLElements | HTMLElementProperties |
                     HTMLElement | string;


function contentToElement(content: EquationInput): HTMLElements {
  if (content instanceof HTMLElements) {
    return content;
  }
  if (content instanceof HTMLElement
      || content instanceof HTMLElementProperties) {
    return new HTMLElements([content]);
  }
  if (typeof content === 'string') {
    return new HTMLElements([new HTMLElement(content)]);
  }

  // Otherwise must be array
  const elementArray: Array<HTMLElement | HTMLElements | HTMLElementProperties | null> = [];

  content.forEach((c) => {
    if (typeof c === 'string') {
      elementArray.push(new HTMLElement(c));
    } else {
      elementArray.push(c);
    }
  });
  return new HTMLElements(elementArray);
}


export default class HTMLEquation extends HTMLElements {
  constructor(id: string = '', classes: string | Array<string> = []) {
    super([], id, classes);
  }

  createEq(content: Array<HTMLElements | HTMLElementProperties | string>) {
    const elements = [];
    content.forEach((c) => {
      if (typeof c === 'string') {
        elements.push(new HTMLElement(c));
      } else {
        elements.push(c);
      }
      this.content = elements;
    });
  }

  htmlElement() {
    const element = document.createElement('div');
    element.setAttribute('id', this.id);
    element.innerHTML = this.render();
    this.classes.forEach((c) => {
      if (c) {
        element.classList.add(c);
      }
    });
    return element;
  }

  // eslint-disable-next-line class-methods-use-this
  el(
    content: 'string',
    id: string = '',
    classes: string | Array<string> = [],
  ) {
    return new HTMLElement(content, id, classes);
  }

  // eslint-disable-next-line class-methods-use-this
  supsub(
    content: EquationInput,
    superscript: EquationInput,
    subscript: EquationInput,
    id: string = '',
    classes: string | Array<string> = [],
  ) {
    return new SuperSub(
      contentToElement(content),
      contentToElement(superscript),
      contentToElement(subscript),
      id,
      classes,
    );
  }

  // eslint-disable-next-line class-methods-use-this
  sub(
    content: EquationInput,
    subscript: EquationInput,
    id: string = '',
    classes: string | Array<string> = [],
  ) {
    return new Subscript(
      contentToElement(content),
      contentToElement(subscript),
      id,
      classes,
    );
  }

  // eslint-disable-next-line class-methods-use-this
  sup(
    content: EquationInput,
    superscript: EquationInput,
    id: string = '',
    classes: string | Array<string> = [],
  ) {
    return new Superscript(
      contentToElement(content),
      contentToElement(superscript),
      id,
      classes,
    );
  }


  // e(
  //   content: string | Array<HTMLElementProperties>,
  //   id: string = '',
  //   classes: string | Array<string> = [],
  // ) {
  //   return new HTMLElements(content, id, classes);
  // }

  // eslint-disable-next-line class-methods-use-this
  frac(
    numerator: EquationInput,
    denominator: EquationInput,
    id: string = '',
    classes: string | Array<string> = [],
  ) {
    return new Fraction(
      contentToElement(numerator),
      contentToElement(denominator),
      id,
      classes,
    );
  }

  // eslint-disable-next-line class-methods-use-this
  sqrt(
    content: EquationInput,
    id: string = '',
    classes: string | Array<string> = [],
  ) {
    return new Root(
      contentToElement(content),
      id,
      classes,
    );
  }
}

