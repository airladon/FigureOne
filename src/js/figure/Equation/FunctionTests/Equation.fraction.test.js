// import {
//   Point,
// } from '../../../../tools/g2';
import {
  round,
} from '../../../tools/math';
import * as tools from '../../../tools/tools';
import makeFigure from '../../../__mocks__/makeFigure';
import { Equation } from '../Equation';
// import Fraction from './Elements/Fraction';

tools.isTouchDevice = jest.fn();

jest.mock('../../Gesture');
jest.mock('../../webgl/webgl');
jest.mock('../../DrawContext2D');

describe('Equation Functions - Fraction', () => {
  let figure;
  let eqn;
  let color1;
  let elements;
  let functions;
  // let forms;
  beforeEach(() => {
    figure = makeFigure();
    color1 = [0.95, 0, 0, 1];
    elements = {
      a: 'a',
      b: 'b',
      c: 'c',
      d: 'd',
      e: 'e',
      f: 'f',
      v: { symbol: 'vinculum' },
      v1: { symbol: 'vinculum' },
    };
    functions = {
      single: () => {
        eqn = new Equation(figure.shapes, { color: color1 });
        const e = eqn.eqn.functions;
        eqn.addElements(elements);
        figure.elements = eqn;
        const s = 0.5;
        eqn.addForms({
          // Full Object
          0: {
            content: {
              frac: {
                numerator: 'a',
                denominator: 'b',
                symbol: 'v',
                scale: s,
                numeratorSpace: 0.01,
                denominatorSpace: 0.02,
                overhang: 0.03,
                offsetY: 0.04,
                baseline: 'vinculum',
                fullContentBounds: false,
              },
            },
          },
          // Method Object
          1: {
            frac: {
              numerator: 'a',
              denominator: 'b',
              symbol: 'v',
              scale: s,
              numeratorSpace: 0.01,
              denominatorSpace: 0.02,
              overhang: 0.03,
              offsetY: 0.04,
              baseline: 'vinculum',
              fullContentBounds: false,
            },
          },
          // Method Array
          2: { frac: ['a', 'v', 'b', s, 0.01, 0.02, 0.03, 0.04, 'vinculum', false] },
          // Function with Method Array
          3: e.frac(['a', 'v', 'b', s, 0.01, 0.02, 0.03, 0.04, 'vinculum', false]),
        });
      },
      scaling: () => {
        eqn = new Equation(figure.shapes, { color: color1 });
        const e = eqn.eqn.functions;
        // const frac = e.frac.bind(e);
        eqn.addElements(elements);
        figure.elements = eqn;
        const s = 0.5;
        eqn.addForms({
          // Method Array with scale defined
          s0: { frac: ['a', 'v', 'b', s] },
          // Function with Method Array with scale defined
          s1: e.frac(['a', 'v', 'b', s]),
          // Bound Function with parameters and scale defined
          // s2: frac('a', 'v', 'b', s),

          // Method Array without scale defined
          n0: { frac: ['a', 'v', 'b'] },
          // Function with Method Array without scale defined
          n1: e.frac(['a', 'v', 'b']),
          // Bound Function with parameters without scale defined
          // n2: frac('a', 'v', 'b'),
        });
      },
      nested: () => {
        eqn = new Equation(figure.shapes, { color: color1 });
        const e = eqn.eqn.functions;
        // const frac = e.frac.bind(e);
        eqn.addElements(elements);
        figure.elements = eqn;
        const s = 0.5;
        eqn.addForms({
          // Full Object
          0: {
            content: {
              frac: {
                numerator: {
                  frac: {
                    numerator: 'a',
                    denominator: 'b',
                    symbol: 'v',
                    scale: s,
                  },
                },
                denominator: 'c',
                symbol: 'v1',
              },
            },
          },
          1: {
            frac: {
              numerator: {
                frac: {
                  numerator: 'a',
                  denominator: 'b',
                  symbol: 'v',
                  scale: s,
                },
              },
              denominator: 'c',
              symbol: 'v1',
            },
          },
          // // Method Array nested in Method Array
          2: {
            frac: [
              {
                frac: ['a', 'v', 'b', s],
              },
              'v1',
              'c',
            ],
          },
          // // Method Array nested in Method Array, all in an Array
          3: [{ frac: [{ frac: ['a', 'v', 'b', s] }, 'v1', 'c'] }],
          // // Method Array in a Function
          4: e.frac([
            {
              frac: ['a', 'v', 'b', s],
            },
            'v1',
            'c',
          ]),
        });
      },
      parameterSteps: () => {
        eqn = new Equation(figure.shapes, { color: color1 });
        // const e = eqn.eqn.functions;
        eqn.addElements(elements);
        figure.elements = eqn;
        eqn.addForms({
          // Full Object
          base: {
            content: {
              frac: {
                numerator: 'a',
                symbol: 'v',
                denominator: 'b',
                scale: 1,
                numeratorSpace: 0.01,
                denominatorSpace: 0.01,
                overhang: 0.01,
                offsetY: 0.01,
              },
            },
            scale: 1,
          },
          numeratorSpace: {
            content: { frac: ['a', 'v', 'b', 1, 0.02, 0.01, 0.01, 0.01] },
            scale: 1,
          },
          denominatorSpace: {
            content: { frac: ['a', 'v', 'b', 1, 0.01, 0.02, 0.01, 0.01] },
            scale: 1,
          },
          overhang: {
            content: { frac: ['a', 'v', 'b', 1, 0.01, 0.01, 0.02, 0.01] },
            scale: 1,
          },
          offsetY: {
            content: { frac: ['a', 'v', 'b', 1, 0.01, 0.01, 0.01, 0.02] },
            scale: 1,
          },
          numeratorBaseline: {
            content: { frac: ['a', 'v', 'b', 1, 0.01, 0.01, 0.01, 0.02, 'numerator'] },
            scale: 1,
          },
          denominatorBaseline: {
            content: { frac: ['a', 'v', 'b', 1, 0.01, 0.01, 0.01, 0.02, 'denominator'] },
            scale: 1,
          },
        });
      },
    };
  });
  describe('Parameter Steps', () => {
    let baseA;
    let baseB;
    let baseV;
    let space;
    let overhang;
    let offsetY;
    beforeEach(() => {
      functions.parameterSteps();
      eqn.showForm('base');
      figure.setFirstTransform();
      baseA = eqn._a.getBoundingRect('figure');
      baseB = eqn._b.getBoundingRect('figure');
      baseV = eqn._v.getBoundingRect('figure');
      space = 0.01;
      overhang = 0.02;
      offsetY = 0.02;
    });
    test('Numerator Space', () => {
      eqn.showForm('numeratorSpace');
      figure.setFirstTransform();
      const newA = eqn._a.getBoundingRect('figure');
      const newB = eqn._b.getBoundingRect('figure');
      const newV = eqn._v.getBoundingRect('figure');
      expect(round(newA.top)).toBe(round(baseA.top + space));
      expect(round(newV.top)).toBe(round(baseV.top));
      expect(round(newB.top)).toBe(round(baseB.top));
    });
    test('Denominator Space', () => {
      eqn.showForm('denominatorSpace');
      figure.setFirstTransform();
      const newA = eqn._a.getBoundingRect('figure');
      const newB = eqn._b.getBoundingRect('figure');
      const newV = eqn._v.getBoundingRect('figure');
      expect(round(newA.top)).toBe(round(baseA.top));
      expect(round(newV.top)).toBe(round(baseV.top));
      expect(round(newB.top)).toBe(round(baseB.top - space));
    });
    test('Overhang', () => {
      eqn.showForm('overhang');
      figure.setFirstTransform();
      const newA = eqn._a.getBoundingRect('figure');
      const newB = eqn._b.getBoundingRect('figure');
      const newV = eqn._v.getBoundingRect('figure');
      expect(round(newA.left)).toBe(round(newV.left + overhang));
      expect(round(newV.left)).toBe(round(baseV.left));
      expect(round(newV.width)).toBe(round(baseV.width + overhang));
      expect(round(newB.left)).toBe(round(newV.left + overhang));
    });
    test('OffsetY', () => {
      eqn.showForm('offsetY');
      figure.setFirstTransform();
      const newV = eqn._v.getBoundingRect('figure');
      expect(round(newV.bottom)).toBe(round(offsetY));
    });
    test('Numerator baseline', () => {
      eqn.showForm('numeratorBaseline');
      figure.setFirstTransform();
      const newA = eqn._a.getBoundingRect('figure');
      const newB = eqn._b.getBoundingRect('figure');
      const newV = eqn._v.getBoundingRect('figure');
      expect(round(newA.bottom)).toBe(-0.008);
      expect(round(newV.top)).toBe(round(newA.bottom - space));
      expect(round(newB.top)).toBe(round(newV.bottom - space));
    });
    test('Denominator baseline', () => {
      eqn.showForm('denominatorBaseline');
      figure.setFirstTransform();
      const newA = eqn._a.getBoundingRect('figure');
      const newB = eqn._b.getBoundingRect('figure');
      const newV = eqn._v.getBoundingRect('figure');
      expect(round(newB.bottom)).toBe(-0.008);
      expect(round(newV.bottom)).toBe(round(newB.top + space));
      expect(round(newA.bottom)).toBe(round(newV.top + space));
    });
  });
  test('Single Fraction', () => {
    functions.single();
    const elems = [eqn._a, eqn._b, eqn._v];
    const formsToTest = ['1', '2', '3'];

    eqn.showForm('0');
    figure.setFirstTransform();
    const positions0 = elems.map(elem => round(elem.transform.mat).slice());
    formsToTest.forEach((f) => {
      eqn.showForm(f);
      figure.setFirstTransform();
      const positions = elems.map(elem => round(elem.transform.mat).slice());
      expect(positions0).toEqual(positions);
    });
  });
  test('Scale Fraction', () => {
    functions.scaling();
    const elems = [eqn._a, eqn._b, eqn._c, eqn._v, eqn._v1];
    const scaleFormsToTest = ['s1'];
    const noScaleFormsToTest = ['n1'];

    // Check all forms that were scaled are the same
    eqn.showForm('s0');
    figure.setFirstTransform();
    let positions0 = elems.map(elem => round(elem.transform.mat).slice());
    scaleFormsToTest.forEach((f) => {
      eqn.showForm(f);
      figure.setFirstTransform();
      const positions = elems.map(elem => round(elem.transform.mat).slice());
      expect(positions0).toEqual(positions);
    });

    // Check all forms that weren't scaled are the same
    eqn.showForm('n0');
    positions0 = elems.map(elem => round(elem.transform.mat).slice());
    noScaleFormsToTest.forEach((f) => {
      eqn.showForm(f);
      const positions = elems.map(elem => round(elem.transform.mat).slice());
      expect(positions0).toEqual(positions);
    });

    // Check height is double the scaled form
    const sHeight = eqn.eqn.forms.s0.content[0].content[0].content[1].height;
    const nHeight = eqn.eqn.forms.n0.content[0].content[0].content[1].height;
    expect(round(nHeight / sHeight)).toBe(2);
  });
  test('Nested Fractions', () => {
    functions.nested();
    const elems = [eqn._a, eqn._b, eqn._c, eqn._v, eqn._v1];
    const formsToTest = ['1', '2', '3', '4', '5'];

    eqn.showForm('0');
    figure.setFirstTransform();
    const positions0 = elems.map(elem => round(elem.transform.mat).slice());
    formsToTest.forEach((f) => {
      eqn.showForm(f);
      figure.setFirstTransform();
      const positions = elems.map(elem => round(elem.transform.mat).slice());
      expect(positions0).toEqual(positions);
    });

    tools.cleanUIDs(eqn);
    expect(round(eqn._a.transform.mat)).toMatchSnapshot();
    expect(round(eqn._b.transform.mat)).toMatchSnapshot();
    expect(round(eqn._c.transform.mat)).toMatchSnapshot();
    expect(round(eqn._v.transform.mat)).toMatchSnapshot();
    expect(round(eqn._v1.transform.mat)).toMatchSnapshot();
  });
});
