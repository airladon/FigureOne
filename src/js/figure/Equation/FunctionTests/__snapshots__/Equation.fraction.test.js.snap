// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Equation Functions - Fraction Nested Fractions 1`] = `
Array [
  0.35,
  0,
  0,
  0.0525,
  0,
  0.35,
  0,
  0.1936,
  0,
  0,
  1,
  0,
  0,
  0,
  0,
  1,
]
`;

exports[`Equation Functions - Fraction Nested Fractions 2`] = `
Array [
  0.35,
  0,
  0,
  0.0525,
  0,
  0.35,
  0,
  0.0968,
  0,
  0,
  1,
  0,
  0,
  0,
  0,
  1,
]
`;

exports[`Equation Functions - Fraction Nested Fractions 3`] = `
Array [
  0.7,
  0,
  0,
  0.035,
  0,
  0.7,
  0,
  -0.0525,
  0,
  0,
  1,
  0,
  0,
  0,
  0,
  1,
]
`;

exports[`Equation Functions - Fraction Nested Fractions 4`] = `
Array [
  0.07,
  0,
  0,
  0.035,
  0,
  0.01,
  0,
  0.1633,
  0,
  0,
  1,
  0,
  0,
  0,
  0,
  1,
]
`;

exports[`Equation Functions - Fraction Nested Fractions 5`] = `
Array [
  0.14,
  0,
  0,
  0,
  0,
  0.01,
  0,
  0.049,
  0,
  0,
  1,
  0,
  0,
  0,
  0,
  1,
]
`;
